File: project-euler-c++/problem81.cpp
// Project Euler Problem 81 (https://projecteuler.net/problem=81)
// July 26, 2020
// Problem already solved in Python; writing C++ solution (without reference to Python solution) 
	// to eventually solve Problems 82 and 83

#include <vector>
#include <string>
#include "auxFunctions.h"
using namespace std;

int minPathSum(int startRow, int startCol, const vector<vector<int>>& matrix, vector<vector<int>>& minPathSums);
	// recursively finds minimal path sum from (startRow, startIndex) to last element

int minPathSumTwoWays(string fileName, int dim) {

	vector<vector<int>> matrix = aux::loadMatrix(fileName, dim);
	vector<vector<int>> minPathSums(dim, vector<int>(dim, -1));
		// create a 2D matrix to store the length of minimal path sums to end point
		// -1 indicates min path sum not found yet
	
	return minPathSum(0, 0, matrix, minPathSums);
}

int minPathSum(int startRow, int startCol, const vector<vector<int>>& matrix, vector<vector<int>>& minPathSums) {
	// recursively finds the minimal path from starting index (startRow, startCol) to last element
		// moving only right and down

	// size of matrix
	const int matrixDim = matrix.size();

	// base case: already at end point
	if (startCol == matrixDim - 1 && startRow == matrixDim - 1)
		return matrix[matrixDim - 1][matrixDim - 1];
	// base case: already found minimal path
	if (minPathSums[startRow][startCol] != -1)
		return minPathSums[startRow][startCol];

	int currentElem = matrix[startRow][startCol];	// current element
	int minSum;	// minimal path sum from current index

	// not at end point -> find minimal path from right element and below element (if they exist)
	// case 1: there exists a path to the right and below
	if (startCol < matrixDim - 1 && startRow < matrixDim - 1) {
		int rightPath = minPathSum(startRow, startCol + 1, matrix, minPathSums);
			// minimal path from the right
		int downPath = minPathSum(startRow + 1, startCol, matrix, minPathSums);
			// minimal path from below
		// finding minimal path
		if (rightPath < downPath)
			minSum = rightPath + currentElem;
		else
			minSum = downPath + currentElem;
	}
	// case 2: there only exists a path to the right
	else if (startCol < matrixDim - 1)
		minSum = minPathSum(startRow, startCol + 1, matrix, minPathSums) + currentElem;
	// case 3: there only exists a path below
	else
		minSum = minPathSum(startRow + 1, startCol, matrix, minPathSums) + currentElem;

	// update minimal paths
	minPathSums[startRow][startCol] = minSum;

	// return minimal path
	return minSum;
}

File: project-euler-julia/problem81.jl
# Julia Solution to Project Euler Problem 81
# 5 August 2020
# Runtime: ~10⁻³ seconds

using DelimitedFiles # readdlm

"""
	minPathSum(sr, sc, matrix, minPathSums)

Recursive function which returns the sum of the minimal path from `matrix[sr, sc]` to `matrix[end, end]`
	going only down and right.

# Arguments
- `sr::Int`: row number of starting element
- `sc::Int`: column number of starting element
- `matrix::Matrix{Int}`: matrix representing a grid of integers
- `minPathSums::Matrix{Union{Int, Nothing}}`: matrix representing minimal path sums such that `minPathSums[i, j]` is the sum of the minimal path from `matrix[i, j]` to `matrix[end, end]` if known, and `nothing` otherwise
"""
function minPathSum(sr::Int, sc::Int, matrix::Matrix{Int},  minPathSums::Matrix{Union{Int, Nothing}})::Int

	# out of bounds, can't travel -> return "infinity"
	!checkbounds(Bool, matrix, sr, sc) && return typemax(Int)	# sentinel for infinity
    # base case: already found minimal path
    minPathSums[sr, sc] ≠ nothing && return minPathSums[sr, sc]

    # not at end point -> find minimal path from right element and below element (if they exist)
		# if they don't exist, line 10 returns infinity
	rightPath = minPathSum(sr, sc + 1, matrix, minPathSums)
	 	# minimal path from the right
	downPath = minPathSum(sr + 1, sc, matrix, minPathSums)
		# minimal path from below

	# finding minimal path and updating minPathSums
	minPathSums[sr, sc] = min(rightPath, downPath) + matrix[sr, sc]

	# returning minimal path
	return minPathSums[sr, sc]

end

"""
	minPathSumTwoWays(fileName)

Returns sum of the minimum path from the top-left element of a grid defined in `fileName` 
to the bottom-right element moving only down and right.
"""
function minPathSumTwoWays(fileName::String)::Int
    # return minimum path sum across matrix defined in fileName moving only down and right

    # extracting matrix
    matrix::Matrix{Int} = readdlm(fileName, ',', Int, '\n')
    # defining minPathSums matrix
    minPathSums = Matrix{Union{Int, Nothing}}(nothing, size(matrix))
		# -1 by default; if minPathSums[i, j] == -1, the minimal path from (1, 1) to (i, j) has not been found

	# setting base case: minimum path from last element to last element is the element itself
	minPathSums[end, end] = matrix[end, end]

	return minPathSum(1, 1, matrix, minPathSums)
end

# function call
@time println(minPathSumTwoWays("Problem Resources\\problems81,82,83.txt"))